### PR TITLE
[makefile] 10x speed up checking/fixing 

### PR DESCRIPTION
**Pull request by stas00**
I present to you an updated `fixup` target which is now super-fast, as it only fixes and validates files that were modified since the branching point, which usually is ~5 out of ~1000. Whoah! Give it a try:

```
make fixup
```
Because of the start up overhead and the 2 customs scripts aren't optimized yet (to check only modified files), currently it's about 10 times faster.

before this PR:
```
time make fixup
real    0m19.272s
user    2m28.253s
sys     0m2.794s
```
after this PR:
```
time make fixup
real    0m2.864s
user    0m2.849s
sys     0m0.778s
```

So what's happening here:

1. `git merge-base --fork-point master` - get a sha of the branching point
2. `git diff --name-only $(git merge-base --fork-point master)` - give us all the filenames that were modified since the branching point (regardless of whether they were staged, and/or pushed or they are still local). The only missing parts would be newly added files that aren't under git yet - shouldn't be a problem though.
3. finally we want to check only specific top-folders so:
`git diff --name-only $(git merge-base --fork-point master) | egrep '^(examples|templates|tests|src|utils)'`
4. now feed that to flake8, black, isort, etc.:
`flake8 $(git diff --name-only $(git merge-base --fork-point master) | egrep '^(examples|templates|tests|src|utils)')`

   but only if there were modified files, so there is an `if` check in the Makefile. otherwise if we get no match - it runs wild on the whole repo and reports things we don't care about.

@sgugger, so if you modify `utils/check_copies.py` to optionally get specific filenames - we can unleash its full power. If you are up for it, do the required modifications and I will take care of the Makefile to pass them on.

----

## refactor repeated dir listings

This PR also refactors repeated dirs  into a single variable on top of the file

@sgugger, @LysandreJik, @sshleifer

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,24 +1,46 @@
-.PHONY: quality_checks quality style fixup test test-examples docs
+.PHONY: modified_only_fixup extra_quality_checks quality style fixup fix-copies test test-examples docs
+
+
+check_dirs := examples templates tests src utils
```
Comment (sshleifer) [inline Makefile:4]: This has always needed to be run from the repo root, right?

**Review by sshleifer — APPROVED** [review]
🔥

File: Makefile (hunk 1 above)
Comment (stas00) [reply Makefile:4]: Yes, since `Makefile` is in root of the repo.

Comment (stas00) [reply Makefile:4]: I haven't changed any behavior, just replaced a multiple duplicated listings with a single variable.

**Comment by sgugger** [comment]
This looks cool! Can add the file argument to `check_copies` since it's easy (you can do it too if you prefer to have it in one PR) but since `check_copies` is super fast now, it probably won't add much.

**Comment by stas00** [comment]
I meant if you wanted to re-enable the blackify functions that were slowing things down.

**Comment by sgugger** [comment]
I'm waiting to see if there actual use cases for that before re-enabling it as there are some ways to make it faster by rewriting the whole script. It would slow down your fast command by quite a bit if a file like roberta has been changed.

**Comment by stas00** [comment]
Hmm, I haven't thought of the scenario of when a branch gets re-based, currently it will add all the files modified since the branching and not just the files modified by the PR. There must be a way to subtract those changes in the master. I will have to think some more.

If we can successfully get that minimal list of files then any developer not working on roberta shouldn't get impacted by its slowdown.

I will follow up after some more experiments.

**Review by LysandreJik — APPROVED** [review]
This is awesome! Thanks @stas00!

**Pull request merged**